packages/react-reconciler/src/ReactChildFiber.js

```diff
@@ -157,7 +156,17 @@ function coerceRef(
         return current.ref;
       }
       const ref = function(value) {
-        const refs = inst.refs === emptyObject ? (inst.refs = {}) : inst.refs;
```

Comment:
This part was a bit shady.

By default we set `inst.refs` to a frozen, shared object. Whenever we want to mutate a ref for the first time, we check whether it's still "the same" pooled object, and if it is, we swap it out with the mutable one.

In some obscure cases where `emptyObject` identity changes over time (e.g. due to a module reset in Jest, or due to a duplicate `fbjs`), this wouldn't work and cause very confusing errors.

I've changed this to use an explicit flag instead. It is only set and read in this module. By default it's not set.

When the flag is not set, the first attempt to set a legacy ref will know that we need to create a new mutable refs object. Then we set the flag. Next time we attach a ref, the flag tells us it's safe to mutate now.

This only affects classes using string refs. There is no cost for classes that don't. Even for those that do, it's a single boolean field.

This lets us get rid of the dependency on strict equality.

Reply:
You’re messing with the hidden classes here so there can still be a cost.

I’m not sure this is worth it.

Reply:
We're already messing with them for legacy context. The only kind of components I'm worried about regressing (Relay) has both legacy context and legacy refs. This gives an incentive to migrate off both without regressing the current situation.

Reply:
I guess context is more localized in practice than legacy refs. Maybe not worth risking.

I can keep it on the React object but I think this will require some mocking of React itself in tests to ensure the object is preserved throughout `resetModules`. It was a bit easier for a separate module.

Reply:
Pushed an alternative fix